src/Tizen.Multimedia.Util/ImageUtil/ImageDecoder.cs

```diff
@@ -204,11 +203,11 @@ private bool CheckHeader(string inputFile)
 
         private IEnumerable<BitmapFrame> RunDecoding()
         {
-            IntPtr outBuffer = IntPtr.Zero;
+            IntPtr outBuffer = Marshal.AllocHGlobal(sizeof(char));
```

Comment:
Is it correct?
outBuffer is used for image decoding, so it seem to need more size than one byte

Reply:
``` c
195 int image_util_decode_set_output_buffer(image_util_decode_h handle, unsigned char **dst_buffer)
196 {
197	decode_s *_handle = (decode_s *) handle;
198
199	image_util_warning("DEPRECATION WARNING: image_util_decode_set_output_buffer() is deprecated and will be removed from next release.");
200
201	image_util_fenter();
202
203	image_util_retvm_if(_handle == NULL, IMAGE_UTIL_ERROR_INVALID_PARAMETER, "Invalid Handle");
204	image_util_retvm_if(dst_buffer == NULL, IMAGE_UTIL_ERROR_INVALID_PARAMETER, "Invalid parameter");
205
206	_handle->dst_buffer = (void **)dst_buffer;
207
208	return IMAGE_UTIL_ERROR_NONE;
209}
```

Reply:
Ok, now I understand why only one byte need, 

``` c#
298int image_util_decode_run(image_util_decode_h handle, unsigned long *width, unsigned long *height, unsigned long long *size)
299{
300	int err = IMAGE_UTIL_ERROR_NONE;
301	decode_s *_handle = (decode_s *) handle;
302	mm_util_image_h image_info = NULL;
303	unsigned int _width = 0;
304	unsigned int _height = 0;
305	unsigned char *_buffer = NULL;
306	size_t _size = 0;
307
308	image_util_warning("DEPRECATION WARNING: image_util_decode_run() is deprecated and will be removed from next release. Use image_util_decode_run2() instead.");
309
310	image_util_retvm_if(_handle == NULL, IMAGE_UTIL_ERROR_INVALID_PARAMETER, "Invalid Handle");
311	image_util_retvm_if((_handle->path == NULL && _handle->src_buffer == NULL), IMAGE_UTIL_ERROR_INVALID_PARAMETER, "Invalid input");
312	image_util_retvm_if(_handle->dst_buffer == NULL, IMAGE_UTIL_ERROR_INVALID_PARAMETER, "Invalid output");
313	image_util_retvm_if(width == NULL || height == NULL, IMAGE_UTIL_ERROR_INVALID_PARAMETER, "Invalid width or height");
314	image_util_retvm_if(size == NULL, IMAGE_UTIL_ERROR_INVALID_PARAMETER, "Invalid size");
315
316	err = __image_util_decode_internal(_handle, &image_info);
317	image_util_retvm_if(err != IMAGE_UTIL_ERROR_NONE, err, "__image_util_decode_internal failed");
318
319	err = mm_image_get_image(image_info, &_width, &_height, NULL, &_buffer, &_size);
320
321	*width = (unsigned long)_width;
322	*height = (unsigned long)_height;
323	*size = (unsigned long long)_size;
324	*(_handle->dst_buffer) = _buffer;
325
326	mm_image_destroy_image(image_info);
327
328	return _image_error_capi(err);
329}
```

But it seem to memory leak the line number 324. 
Who is release the _buffer?

Reply:
It's released by following code.
```
finally
{
    Marshal.FreeHGlobal(outBuffer);
}
```

Reply:
it just release a one byte that allocated by Marshal.AllocHGlobal(sizeof(char));

you need to release a memory that is indicated by *outBuffer

Reply:
``` c
39/* After @data is used, it must be released using free().*/
40 int mm_image_get_image(mm_util_image_h image, unsigned int *width,
41		unsigned int *height, mm_util_color_format_e *color,
42		unsigned char **data, size_t *size);
```

Reply:
According to your comment, the following code should be working well, right?

```c
int image_util_free(image_util_decode_h handle)
{
	int err = IMAGE_UTIL_ERROR_NONE;
	decode_s *_handle = (decode_s *) handle;

	...

	image_util_warning("dst_buffer : %p,
            *buffer : %p", _handle->dst_buffer,  *(_handle->dst_buffer));

	free(*(_handle->dst_buffer));

	return _image_error_capi(err);
}
```

```csharp
[DllImport(Libraries.ImageUtil, EntryPoint = "image_util_free")]
internal static extern ImageUtilError FreeNativeBuffer(ImageDecoderHandle handle);

private IEnumerable<BitmapFrame> RunDecoding()
{
	IntPtr outBuffer = Marshal.AllocHGlobal(sizeof(char));

	try
	{
		SetOutputBuffer(Handle, ref outBuffer).
                        ThrowIfFailed("Failed to decode given image");

		DecodeRun(Handle, out var width, out var height, out var size).
                        ThrowIfFailed("Failed to decode");

		yield return new BitmapFrame(outBuffer, width, height, (int)size);
	}
	finally
	{
		FreeNativeBuffer(Handle);

		Marshal.FreeHGlobal(outBuffer);
	}
}
```

Reply:
Yes, it will be working well

Reply:
But, unfortunately, It make crash - double free or corruption.
It seems that `Marshal.FreeHGlobal(outBuffer)` try to free not `**outBuffer(0xacc40848)` but `*outBuffer (0xb7b724a8)`.



07-04 18:49:35.164+0900 W/CAPI_MEDIA_IMAGE_UTIL(P20906, T20906): image_util_decode.c: image_util_decode_set_output_buffer(208) > **dst_buffer addr : **0xacc40848**

07-04 18:49:35.194+0900 W/CAPI_MEDIA_IMAGE_UTIL(P20906, T20906): image_util_decode.c: image_util_decode_run(322) > _buffer address : 0xb7b724a8
07-04 18:49:35.194+0900 W/CAPI_MEDIA_IMAGE_UTIL(P20906, T20906): image_util_decode.c: image_util_decode_run(329) > *(_handle->dst_buffer) addr : 0xb7b724a8

07-04 18:49:35.194+0900 I/Tizen.Multimedia.Util(P20906, T20906): BitmapFrame.cs: .ctor(33) > Enter

07-04 18:49:35.204+0900 I/Tizen.Multimedia.Util(P20906, T20906): ImageDecoder.cs: RunDecoding(223) > Call FreeBuffer

07-04 18:49:35.214+0900 W/CAPI_MEDIA_IMAGE_UTIL(P20906, T20906): image_util_decode.c: image_util_free(344) > dst_buffer : 0xacc40848, *buffer: 0xb7b724a8
07-04 18:49:35.214+0900 W/CAPI_MEDIA_IMAGE_UTIL(P20906, T20906): image_util_decode.c: image_util_free(347) > free buffer done

07-04 18:49:35.214+0900 I/Tizen.Multimedia.Util(P20906, T20906): ImageDecoder.cs: RunDecoding(226) > Call Marshal.FreeHGlobal

07-04 18:49:35.214+0900 I/DOTNET_LAUNCHER(P20906, T20928): *** Error in `/opt/usr/globalapps/Tizen.MultimediaUtil.Tests/bin/Tizen.MultimediaUtil.Tests.dll': double free or corruption (top): **0xb7b724a8** ***

Reply:
First, 
``` c#
        [DllImport(Libraries.ImageUtil, EntryPoint = "image_util_decode_set_output_buffer")]
        internal static extern ImageUtilError SetOutputBuffer(ImageDecoderHandle handle, ref IntPtr dstBuffer);


IntPtr outBuffer = Marshal.AllocHGlobal(sizeof(char));
SetOutputBuffer(Handle, ref outBuffer);
```
it is something wrong. 
`SetOutputBuffer(Handle, ref outBuffer);`
It will pass a address of outBuffer of address instead of outBuffer
You should pass a memory address that assigned by Marshal.AllocHGlobal

image_util_decode_set_output_buffer API design was wrong. 
Why do you deprecated API?

Reply:
It will be working correctly, I have not test envirement, please test it
``` c#
private IEnumerable<BitmapFrame> RunDecoding()
{
	IntPtr outBuffer = IntPtr.Zero;

	try
	{
		SetOutputBuffer(Handle, ref outBuffer).
                        ThrowIfFailed("Failed to decode given image");

		DecodeRun(Handle, out var width, out var height, out var size).
                        ThrowIfFailed("Failed to decode");

		yield return new BitmapFrame(outBuffer, width, height, (int)size);
	}
	finally
	{
		FreeNativeBuffer(Handle);
	}
}
```

Reply:
hm... it is the orignal code. Why do you submit PR?

Reply:
`outBuffer` doesn't hold address, even if Native FW set buffer to it correctly.

That is, 
1. SetOutBuffer() => hold the address by native fw of `outBuffer`
2. DecodeRun()   => set buffer to `outBuffer` correctly.
3. Sometimes, the address of `outBuffer`  is `0`

I thought that original `outBuffer` was in manage memory and, its address can be moved by GC between step 1. ~ step 2 as you know.
So I allocated the `outBuffer` into unmanaged memory and it's working very well.

Reply:
Ok, manage memory (IntPtr outBuffer) is moved when GC was occurred, 
So, it should be pinned on the some memory area and AllocHGlobal pinning a memory address. 
but current code is wrong 
It still use unpinned memory address

It should be fix like below

``` c#
        [DllImport(Libraries.ImageUtil, EntryPoint = "image_util_decode_set_output_buffer")]
        internal static extern ImageUtilError SetOutputBuffer(ImageDecoderHandle handle, IntPtr dstBuffer);

        private IEnumerable<BitmapFrame> RunDecoding()
        {
            IntPtr outBuffer = Marshal.AllocHGlobal(sizeof(char));

            try
            {
                SetOutputBuffer(Handle, outBuffer).ThrowIfFailed("Failed to decode given image");

                DecodeRun(Handle, out var width, out var height, out var size).
                    ThrowIfFailed("Failed to decode");

                yield return new BitmapFrame(outBuffer, width, height, (int)size);
            }
            finally
            {
                FreeNativeBuffer(Handle);
                Marshal.FreeHGlobal(outBuffer);
            }
        }

```

Reply:
I've tested both two size : `sizeof(char)` and `Marshal.SizeOf(typeof(IntPtr))`
It's working well both. but I agree that `Marshal.SizeOf()` is better solution.
I'll update code.

And the answer about `Why do you use deprecated API?` is that,
image_util_decode_set_output_buffer() and image_util_decode_run() is used together.
the replacement API of these are introduced in Tizen 5.5, but Crash issue is occurred in Tizen 5.0 also. so we can't use replacement API. 
I'll fix the bug of all tizen version with this patch and I'll try to replace Level 6 P/Invoke function to replacement API.

Reply:
>I've tested both two size : sizeof(char) and Marshal.SizeOf(typeof(IntPtr))
>It's working well both. but I agree that Marshal.SizeOf() is better solution.
Because dotnet core runtime is allocating a bulk of memory chunk. if you use sizeof(char), it will access memory area that allocated by others

Reply:
OK I see. Thank you for the review.

Reply:
And.. 
``` C# 
yield return new BitmapFrame(outBuffer, width, height, (int)size);
yield return new BitmapFrame(Marshal.ReadIntPtr(outBuffer), width, height, (int)size);
```
We also need to update it, isn't it?